doc/port_link_training/port-link-training-design.md

```diff
@@ -0,0 +1,336 @@
+
+| **Parameter**          | **sai_port_attr_t**
+|:-----------------------|:-------------------------------------------|
+| link_training          | SAI_PORT_ATTR_LINK_TRAINING_ENABLE         |
+| link_training_failure  | SAI_PORT_ATTR_LINK_TRAINING_FAILURE_STATUS |
+| link_training_rxstatus | SAI_PORT_ATTR_LINK_TRAINING_RX_STATUS      |
+
+Please note the operational port link training status in the APPL_DB could be different from the administratively configured link training configuration in the CONFIG_DB due to the hardware limitations or vendor-specific SAI software implementations. For example, port link training is only applicable to the ports with speed >= 10G, and it's not applicable for chip-to-module transceivers.
+
+#### YANG Model Enhancements
+
+The port yang model needs to update according to DB schema change. The yang model changes of new fields are described below:
+
+```
+leaf link_training {
+  type string {
+    pattern "auto|on|off";
+  }
+}
+```
+
+These new yang leaf will be added to sonic-port.yang.
+
+#### DB Migration Considerations
+
+By having port link training disabled if "link_training" is not specified in the CONFIG_DB, this feature is fully backward compatible.
+
+#### SWSS Enhancements
+
+##### Port Configuration Flow
+
+The current SONiC port configuration flow in PortsOrch can be described in following pseudo code:
+
+```
+port = getPort(alias)
+if autoneg changed:
+    setPortAutoNeg(port, autoneg)
+```
+
+The new SONiC port configuration flow can be described in following pseudo code:
+
+```
+port = getPort(alias)
+if autoneg changed:
+    setPortAutoNeg(port, autoneg)
+
+if link_training changed and xcvr_capabilities has LT:
+    setPortLinkTraining(port, link_training)
+```
+
+##### Link Training Status Refresh
+
+To refresh the operational link training status in APPL_DB, a new notification handler will be introduced to PortsOrch, the pseudo code is as follow:
+
+```
+portCtrlConsumer = NotificationConsumer("PORT_CTRL")
+if portCtrlConsumer received 'op=port_status_refresh', 'data=EthernetXY' and 'type=link_training':
+    refreshPortLinkTrainingStatus(EthernetXY)
+```
+
+An example code snippet of initiating status refresh is as follow:
+
+```
+appl_db = daemon_base.db_connect("APPL_DB")
+port_np = swsscommon.NotificationProducer(appl_db, "PORT_CTRL")
+
+fvp = swsscommon.FieldValuePairs([
+    ('type', 'link_training')
+])
+port_np.send('port_state_refresh', port, fvp)
+```
+
+#### PMON xcvrd Considerations
+
+Given that port link training shall not be enabled on certain transceivers. For example, a chip-to-module transceiver. It's better to have pmon#xcvrd provide a hint to the swss#orchagent for the advanced link training controls. Hence, **xcvr_capabilities** is introduced into APPL_DB.  
+
+When **link_training=auto**, swss#orchagent should request syncd to enable port link training only if **LT** is specified in **xcvr_capabilities**.
+
+### Warmboot Design Impact
```

Comment:
Please add considerations for fast reboot. Currently pmon starts after syncd and with link training in place how will this be impacted?

Reply:
If I'm right about it, the fast-reboot is a case that kernel will be restarted from loading a new ELF file, without BIOS intervention, and this will be a COLD-REBOOT scenario, from the switch silicon, swss and pmon perspective, hence none extra care will be necessary.
In the case of warm-reboot, the APPL_DB and STATE_DB will be restored, and the switch silicon hardware state will not be changed and seamless resumed, all we need is to make sure static pre-emphasis request will not be applied in this scenario. Of course, a system test could provide more details, and I'll update the HLD if any issues are observed.

Reply:
When I meant in case of fast-boot, the requirement is to have downtime less than 30 seconds. In normal scenario this wouldn't be a problem as syncd can come up and initialize the ports whereas in this case the xcvrd should publish the capabilities which will be required for link to be up. So the link downtime may be greater than 30 seconds. Please verify this scenario

Reply:
Thanks for the comments, and based on the conclusion of the earlier meeting, the changes against xcvrd will be dropped, and the config command should directly reference the TRANSCEIVER_INFO of STATE_DB for the ability checks.